src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java

```diff
@@ -744,71 +756,102 @@ public void apply(final BiConsumer<String, byte[]> outputConsumer, InputTag... i
-							entry.setValue(data);
-						} else {
-							outputConsumer.accept(mapClass(cls.getName()), data);
+						for (Map.Entry<ClassInstance, byte[]> entry : outputBuffer.entrySet()) {
+							ClassInstance cls = entry.getKey();
+							byte[] data = entry.getValue();
+
+							if (needsFixes) {
+								data = fixClass(cls, data);
+							}
+
+							if (hasInputTags) {
+								entry.setValue(data);
+							} else {
+								outputConsumer.accept(ClassInstance.getMrjName(mapClass(cls.getName()), cls.getMrjVersion()), data);
+							}
 						}
+
+						if (!hasInputTags) outputBuffer = null; // don't expect repeat invocations
+
+						classesToMakePublic.clear();
+						membersToMakePublic.clear();
+					} else if (needsFixes) {
+						throw new RuntimeException(String.format("%d classes and %d members need access fixes", classesToMakePublic.size(), membersToMakePublic.size()));
 					}
+				}
 
-					if (!hasInputTags) outputBuffer = null; // don't expect repeat invocations
+				assert hasInputTags == (outputBuffer != null);
 
-					classesToMakePublic.clear();
-					membersToMakePublic.clear();
-				} else if (needsFixes) {
-					throw new RuntimeException(String.format("%d classes and %d members need access fixes", classesToMakePublic.size(), membersToMakePublic.size()));
+				if (outputBuffer != null) { // partial output selected by input tags
+					for (Map.Entry<ClassInstance, byte[]> entry : outputBuffer.entrySet()) {
+						ClassInstance cls = entry.getKey();
+
+						if (inputTags == null || cls.hasAnyInputTag(inputTags)) {
+							outputConsumer.accept(ClassInstance.getMrjName(mapClass(cls.getName()), cls.getMrjVersion()), entry.getValue());
+						}
+					}
 				}
 			}
 		}
+	}
 
-		assert hasInputTags == (outputBuffer != null);
-
-		if (outputBuffer != null) { // partial output selected by input tags
-			for (Map.Entry<ClassInstance, byte[]> entry : outputBuffer.entrySet()) {
-				ClassInstance cls = entry.getKey();
+	/**
+	 * This function will setup {@code mrjClasses} with any new MRJ version
+	 * added. It will put the result of {@code constructMrjCopy} from lower
+	 * MRJ version to the new version.
+	 * @param versions the new versions that need to be added in to {@code mrjClasses}
+	 */
+	private void fixMrjClasses(Set<Integer> versions) {
+		// ensure the new version is added from lowest to highest
+		for (int toVersion: versions.stream().sorted().collect(Collectors.toList())) {
+			Map<String, ClassInstance> toClasses = new HashMap<>();
+
+			if(mrjClasses.put(toVersion, toClasses) != null) {
+				throw new RuntimeException("internal error: duplicate versions in mrjClasses");
+			}
 
-				if (inputTags == null || cls.hasAnyInputTag(inputTags)) {
-					outputConsumer.accept(mapClass(cls.getName()), entry.getValue());
+			// find the fromVersion that just lower the the toVersion
+			Optional<Integer> fromVersion = mrjClasses.keySet().stream()
+					.filter(v -> v < toVersion).max(Integer::compare);
```

Comment:
I think you can do that cleaner by rememberint the toVersion of the previous version loop iteration.

Reply:
Emm. I do not see how that would be cleaner, and I don't understand how you want to change it. Maybe you could change the code on GitHub. As long as the functionalities are the same, I am fine with it.

Reply:
I tried to PR to your branch, but GH failed. I meant like so, assuming its logic is accurate:

```java
private void fixMrjClasses(Set<Integer> versions) {
	int lastVersion = 0;

	// ensure the new version is added from lowest to highest
	for (int toVersion: versions.stream().sorted().collect(Collectors.toList())) {
		Map<String, ClassInstance> toClasses = new HashMap<>();

		if (mrjClasses.put(toVersion, toClasses) != null) {
			throw new RuntimeException("internal error: duplicate versions in mrjClasses");
		}

		if (lastVersion > 0) {
			Map<String, ClassInstance> fromClasses = mrjClasses.get(lastVersion);

			for (ClassInstance cls: fromClasses.values()) {
				addClass(cls.constructMrjCopy(), toClasses, false);
			}
		}

		lastVersion = toVersion;
	}
}
```

Reply:
This code may not work properly. Suppose `mrjClasses` have version `8` and `9`, and `versions` contain `11`. Then the loop will only run once and therefore old classes won't be added to the `mrjClasses`.
The difference is that your code is trying to find an appreciated version from `toVersion` (i.e. `versions`), where my code is trying to find appreciate version from `mrjClasses`. The latter one will contain more versions than `toVersion`.

Reply:
You mean like when letting TR run, add another input, then run it again?

Reply:
yes.

Reply:
Good to know, not sure it is safe to not start over with the mrj copies in every case though, which is why I didn't expect this.

Reply:
Just realize the current code still have the possibility to have the wrong MRJ resolution. Let me fix that.

Reply:
fixed